lib/galaxy/webapps/galaxy/api/jobs.py

```diff
@@ -257,148 +266,262 @@ def common_problems(self, trans: ProvidesUserContext, id, **kwd):
+        trans: ProvidesUserContext = DependsOnTrans,
+    ) -> List[Optional[JobMetric]]:
+        job = self.service.get_job(trans, job_id=id, hda_ldda=hda_ldda)
+        return [JobMetric(**metric) for metric in summarize_job_metrics(trans, job)]
+
+    @router.get(
+        "/api/datasets/{id}/metrics",
+        name="get_metrics",
+        summary="Return job metrics for specified job.",
+    )
+    def metrics_by_dataset(
+        self,
+        id: Annotated[DecodedDatabaseIdField, DatasetIdPathParam],
+        hda_ldda: Annotated[DatasetSourceType, HdaLddaQueryParam] = DatasetSourceType.hda,
+        trans: ProvidesUserContext = DependsOnTrans,
+    ) -> List[Optional[JobMetric]]:
+        job = self.service.get_job(trans, dataset_id=id, hda_ldda=hda_ldda)
+        return [JobMetric(**metric) for metric in summarize_job_metrics(trans, job)]
+
+    @router.get(
+        "/api/jobs/{job_id}/destination_params",
+        name="destination_params_job",
+        summary="Return destination parameters for specified job.",
+        require_admin=True,
+    )
+    def destination_params(
+        self,
+        job_id: Annotated[DecodedDatabaseIdField, JobIdPathParam],
+        trans: ProvidesUserContext = DependsOnTrans,
+    ) -> JobDestinationParams:
+        job = self.service.get_job(trans, job_id=job_id)
+        return JobDestinationParams(**summarize_destination_params(trans, job))
+
+    @router.post(
+        "/api/jobs/search",
+        name="search_jobs",
+        summary="Return jobs for current user",
+    )
+    def search(
+        self,
+        payload: Annotated[SearchJobsPayload, SearchJobBody],
+        trans: ProvidesHistoryContext = DependsOnTrans,
+    ) -> List[EncodedJobDetails]:
         """
-        * GET /api/jobs/{job_id}/parameters_display
-        * GET /api/datasets/{dataset_id}/parameters_display
-
-            Resolve parameters as a list for nested display. More client logic
-            here than is ideal but it is hard to reason about tool parameter
-            types on the client relative to the server. Job accessibility checks
-            are slightly different than dataset checks, so both methods are
-            available.
+        This method is designed to scan the list of previously run jobs and find records of jobs that had
+        the exact some input parameters and datasets. This can be used to minimize the amount of repeated work, and simply
+        recycle the old results.
+        """
+        tool_id = payload.tool_id
 
-            This API endpoint is unstable and tied heavily to Galaxy's JS client code,
-            this endpoint will change frequently.
+        tool = trans.app.toolbox.get_tool(tool_id)
+        if tool is None:
+            raise exceptions.ObjectNotFound("Requested tool not found")
+        # TODO the inputs are actually a dict, but are passed as a JSON dump
+        # maybe change it?
+        inputs = json.loads(payload.inputs)
```

Comment:
The inputs are passed as a JSON string in the payload. Can I somehow decode the JSON string inside the pydantic model and then typecheck it before calling the actual endpoint logic?
This would produce a cleaner error or am I wrong?

Reply:
Thank you for implementing this!